compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/FunctionCodegen.kt

```diff
@@ -62,14 +62,11 @@ open class FunctionCodegen(
         }
 
         // FIXME: The following test is a workaround for a bug in anonymous object regeneration.
-        //        We currently need to avoid parameter annotations on the (synthetic) constructors of inlined anonymous objects,
+        //        We currently need to avoid parameter annotations on the constructors of inlined anonymous objects,
         //        since otherwise anonymous object regeneration can fail with an ArrayIndexOutOfBounds exception if the number
-        //        or arguments to the constructor changes.
-        if (!hasSyntheticFlag ||
-            irFunction.origin == JvmLoweredDeclarationOrigin.SYNTHETIC_METHOD_FOR_PROPERTY_ANNOTATIONS ||
-            //TODO: investigate this case: annotation here is generated twice in lowered function and in interface method overload
-            irFunction.origin == JvmLoweredDeclarationOrigin.GENERATED_SAM_IMPLEMENTATION
-        ) {
+        //        of arguments to the constructor changes. This hack can be removed after at least one version is released
+        //        that includes `ShiftParameterAnnotationsVisitor`.
+        if (irFunction !is IrConstructor || !(hasSyntheticFlag || irFunction.parentAsClass.isAnonymousObject)) {
```

Comment:
@pyos Actually there shouldn't be any annotation on anonymous object constructor, so looks like after adding `irFunction.parentAsClass.isAnonymousObject` all other changes in first 'Shift parameter annotations on constructors of local classes pyos' commit are redundant  (sorry if I missed smth in previous pr discussions)

Reply:
Yeah, now that I think about it, it really doesn't make sense to support them in the inliner.

Reply:
As this commit unmutes a bunch of tests I could merge it if you extract it from this pr

Reply:
Extracted to #2681